util.c

```diff
@@ -608,151 +571,111 @@ static int parse_ssh_format(const cfg_t *cfg, char *buf, size_t buf_size,
+             memcmp(str, SSH_EDDSA, SSH_EDDSA_LEN) == 0) {
+    type = COSE_EDDSA;
+    point_len = SSH_EDDSA_POINT_LEN;
   } else {
     if (cfg->debug) {
-      D(cfg->debug_file, "Unknown key type %s", decoded);
+      D(cfg->debug_file, "Unknown key type %s", str);
     }
     goto out;
   }
 
-  decoded += len;
-  decoded_len -= len;
-
-  // curve name
-  if (decoded_len < sizeof(uint32_t)) {
-    if (cfg->debug) {
-      D(cfg->debug_file, "Malformed SSH key (curvename length)");
-    }
-    goto out;
-  }
-  len = ntohl(*((uint32_t *) decoded));
-  decoded += sizeof(uint32_t);
-  decoded_len -= sizeof(uint32_t);
-
-  if (decoded_len < len) {
-    if (cfg->debug) {
-      D(cfg->debug_file, "Malformed SSH key (curvename)");
-    }
-    goto out;
+  if (cfg->debug) {
+    D(cfg->debug_file, "keytype (%" PRIu32 ") \"%s\"", len, str);
   }
 
-  // TODO(adma): Add support for eddsa
-  if (len == SSH_P256_NAME_LEN &&
-      memcmp(decoded, SSH_P256_NAME, SSH_P256_NAME_LEN) == 0) {
-    if (cfg->debug) {
-      D(cfg->debug_file, "curvename (%u) \"%s\"", len, decoded);
-    }
-  } else {
-    if (cfg->debug) {
-      D(cfg->debug_file, "Unknown curve %s", decoded);
+  if (type == COSE_ES256) {
+    // curve name
+    if (!ssh_read_str(&decoded, &decoded_len, &str, &len)) {
+      if (cfg->debug) {
+        D(cfg->debug_file, "Malformed SSH key (curvename)");
+      }
+      goto out;
     }
-    goto out;
-  }
 
-  devices[0].coseType = strdup("es256");
-  if (devices[0].coseType == NULL) {
-    if (cfg->debug) {
-      D(cfg->debug_file, "Unable to allocate COSE type");
+    if (len == SSH_P256_NAME_LEN &&
+        memcmp(str, SSH_P256_NAME, SSH_P256_NAME_LEN) == 0) {
+      if (cfg->debug) {
+        D(cfg->debug_file, "curvename (%" PRIu32 ") \"%s\"", len, str);
+      }
+    } else {
+      if (cfg->debug) {
+        D(cfg->debug_file, "Unknown curve %s", str);
+      }
+      goto out;
     }
-    goto out;
   }
 
-  decoded += len;
-  decoded_len -= len;
-
   // point
-  if (decoded_len < sizeof(uint32_t)) {
-    if (cfg->debug) {
-      D(cfg->debug_file, "Malformed SSH key (point length)");
-    }
-    goto out;
-  }
-  len = ntohl(*((uint32_t *) decoded));
-  decoded += sizeof(uint32_t);
-  decoded_len -= sizeof(uint32_t);
-
-  if (decoded_len < len) {
+  if (!ssh_read_str(&decoded, &decoded_len, &str, &len)) {
     if (cfg->debug) {
       D(cfg->debug_file, "Malformed SSH key (point)");
     }
     goto out;
   }
 
-  if (len != 65) { // TODO(adma): unmagify and add support for eddsa
+  if (len != point_len) {
     if (cfg->debug) {
-      D(cfg->debug_file, "Invalid point length, should be %d, found %d", 65,
-        len);
+      D(cfg->debug_file,
+        "Invalid point length, should be %" PRIu32 ", found %" PRIu32 "",
+        point_len, len);
     }
     goto out;
   }
 
   // Skip the initial '04'
-  if (!b64_encode(decoded + 1, len - 1, &devices[0].publicKey)) {
+  if (type == COSE_ES256) {
+    str++;
+    len--;
```

Comment:
i understand this might not be the case with the code as it stands, but i recommend checking if `len == 0` before decrementing it

Reply:
The rebase on top of #201 adds a check for this;
https://github.com/Yubico/pam-u2f/blob/4d5030fe103949241162090c359c3994b752271a/util.c#L589-L594